tests/drivers/regulator/fixed/src/main.c

```diff
@@ -0,0 +1,308 @@
+	}
+	if (rc == 0) {
+		rc = reg_status();
+	}
+
+	if (rc != 0) {		/* should be inactive */
+		precheck = PC_FAIL_UNCONFIGURE;
+		return -EIO;
+	}
+
+	precheck = PC_OK;
+
+	return rc;
+}
+
+/* The regulator driver initializes in POST_KERNEL since it has
+ * thread-related stuff in it.  We need to verify the shorted signals
+ * required by the test before the driver configures its GPIO.  This
+ * should be done late PRE_KERNEL_9, but it can't because Nordic and
+ * possibly other systems initialize GPIO drivers post-kernel. */
+BUILD_ASSERT(CONFIG_REGULATOR_FIXED_INIT_PRIORITY > 74);
+SYS_INIT(setup, POST_KERNEL, 74);
+
+static void test_preconditions(void)
+{
+	zassert_equal(precheck, PC_OK,
+		      "precheck failed: %s",
+		      pc_errstr[precheck]);
+
+	zassert_not_equal(reg_dev, NULL,
+			  "no regulator device");
+}
+
+static void test_basic(void)
+{
+	zassert_equal(precheck, PC_OK,
+		      "precheck failed: %s",
+		      pc_errstr[precheck]);
+
+	zassert_not_equal(reg_dev, NULL,
+			  "no regulator device");
+
+	int rs = reg_status();
+
+	/* Initial state: on if and only if it's always on or was enabled at
+	 * boot.
+	 */
+
+	if (IS_ENABLED(BOOT_ON) || IS_ENABLED(ALWAYS_ON)) {
+		zassert_equal(rs, 1,
+			      "not on at boot: %d", rs);
+	} else {
+		zassert_equal(rs, 0,
+			      "not off at boot: %d", rs);
+	}
+
+	reset_client();
+
+	/* Turn it on */
+	int rc = regulator_enable(reg_dev, &cli);
+	zassert_true(rc >= 0,
+		     "first enable failed: %d", rc);
+
+	if (STARTUP_DELAY_US > 0) {
+		rc = sys_notify_fetch_result(&cli.notify, &rc);
+
+		zassert_equal(rc, -EAGAIN,
+			      "startup notify early: %d", rc);
+
+		while (sys_notify_fetch_result(&cli.notify, &rc) == -EAGAIN) {
+			k_yield();
+		}
+	}
+
+	zassert_equal(callback_cli, &cli,
+		      "callback not invoked");
+	zassert_equal(callback_res, 0,
+		      "callback res: %d", callback_res);
+	zassert_equal(callback_state, ONOFF_STATE_ON,
+		      "callback state: 0x%x", callback_res);
+
+	/* Make sure it's on */
+
+	rs = reg_status();
+	zassert_equal(rs, 1,
+		      "bad on state: %d", rs);
+
+	/* Turn it on again (another client) */
+
+	reset_client();
+	rc = regulator_enable(reg_dev, &cli);
+	zassert_true(rc >= 0,
+		     "second enable failed: %d", rc);
+
+	zassert_equal(callback_cli, &cli,
+		      "callback not invoked");
+	zassert_true(callback_res >= 0,
+		      "callback res: %d", callback_res);
+	zassert_equal(callback_state, ONOFF_STATE_ON,
+		      "callback state: 0x%x", callback_res);
+
+	/* Make sure it's still on */
+
+	rs = reg_status();
+	zassert_equal(rs, 1,
+		      "bad 2x on state: %d", rs);
+
+	/* Turn it off once (still has a client) */
+
+	rc = regulator_disable(reg_dev);
+	zassert_true(rc >= 0,
+		     "first disable failed: %d", rc);
+
+	/* Make sure it's still on */
+
+	rs = reg_status();
+	zassert_equal(rs, 1,
+		      "bad 2x on 1x off state: %d", rs);
+
+	/* Turn it off again (no more clients) */
+
```

Comment:
I feel that having blank lines after each comment is a big much. It's a matter of taste I suppose. I'll get a taller monitor :-)